source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,194 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+                    &thread_local_dispatcher = cluster_manager.thread_local_dispatcher_] {
+    // Check for the cluster here too. It might have been added between the time when this closure
+    // was posted and when it is being executed.
+    if (getThreadLocalCluster(name) != nullptr) {
+      if (weak_callback.expired()) {
+        // Not only the cluster was added, but it was also already handled, so don't bother with
+        // posting the callback back to the worker thread.
+        return;
+      }
+      thread_local_dispatcher.post([weak_callback] {
+        if (auto callback = weak_callback.lock(); callback != nullptr) {
+          // If this gets called here, it means that we requested a discovery of a cluster without
+          // checking if that cluster is already known by cluster manager.
+          (*callback)(ClusterDiscoveryStatus::Available);
+        }
+      });
+      return;
+    }
+
+    if (auto it = pending_cluster_creations_.find(name); it != pending_cluster_creations_.end()) {
+      // We already began the discovery process for this cluster, nothing to do. If we got here, it
+      // means that it was other worker thread that requested the discovery.
+      return;
+    }
+    auto& odcds = odcds_handle->getOdCds();
+    // Start the discovery. If the cluster gets discovered, cluster manager will warm it up and
+    // invoke the cluster lifecycle callbacks, that will in turn invoke our callback.
+    odcds.updateOnDemand(name);
+    // Setup the discovery timeout timer to avoid keeping callbacks indefinitely.
+    auto timer_cb = Event::TimerCb([this, name] { notifyExpiredDiscovery(name); });
+    auto timer = dispatcher_.createTimer(timer_cb);
+    // TODO(krnowak): This timeout period is arbitrary. Should it be a parameter or do we keep it,
+    // but rather with a name?
+    timer->enableTimer(std::chrono::milliseconds(5000));
+    // Keep odcds handle alive for the duration of the discovery process.
+    pending_cluster_creations_.insert(
+        {std::move(name), ClusterCreation{std::move(odcds_handle), std::move(timer)}});
+  });
+
+  return std::move(handle);
+}
+
+void ClusterManagerImpl::notifyExpiredDiscovery(const std::string& name) {
+  auto map_node_handle = pending_cluster_creations_.extract(name);
+  if (map_node_handle.empty()) {
+    return;
+  }
+  // Defer destroying the timer, so it's not destroyed during its callback. TimerPtr is a
+  // unique_ptr, which is not copyable, but std::function is copyable, we turn a move-only
+  // unique_ptr into a copyable shared_ptr and pass that to the std::function.
+  dispatcher_.post([timer = std::shared_ptr<Event::Timer>(
+                        std::move(map_node_handle.mapped().expiration_timer_))] {});
+
+  // Let all the worker threads know that the discovery timed out.
+  tls_.runOnAllThreads([name](OptRef<ThreadLocalClusterManagerImpl> cluster_manager) {
+    if (!cluster_manager.has_value()) {
+      return;
+    }
```

Comment:
I don't think this is possible?

Reply:
It's an opt-ref so I thought that the check would be prudent to do. I'll remove it.